workflow/taskset/task_set_manager.go

```diff
@@ -0,0 +1,166 @@
+	return &WorkflowTaskSetManager{
+		wfTaskSetClient:   client,
+		wfTaskSetInformer: informer,
+		queueWorkflowFunc: queueFunc,
+		wfTaskSetQueue:    metrics.RateLimiterWithBusyWorkers(workqueue.DefaultControllerRateLimiter(), "wf_task_set_queue"),
+	}
+}
+
+func (wfts WorkflowTaskSetManager) CreateTaskSet(ctx context.Context, wf *wfv1.Workflow, nodeId string, tmpl wfv1.Template) error {
+	key := fmt.Sprintf("%s/%s", wf.Namespace, wf.Name)
+	log.WithField("workflow", wf.Name).WithField("namespace", wf.Namespace).WithField("TaskSet", key).Infof("Creating TaskSet")
+
+	obj, exists, err := wfts.wfTaskSetInformer.Informer().GetIndexer().GetByKey(key)
+	if err != nil {
+		log.WithError(err).Error(fmt.Sprintf("Failed to get TaskSet '%s' from informer index", key))
+		return err
+	}
+
+	if !exists {
+		taskSet := wfv1.WorkflowTaskSet{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       workflow.WorkflowTaskSetKind,
+				APIVersion: workflow.APIVersion,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: wf.Namespace,
+				Name:      wf.Name,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion: wf.APIVersion,
+						Kind:       wf.Kind,
+						UID:        wf.UID,
+						Name:       wf.Name,
+					},
+				},
+			},
+			Spec: wfv1.WorkflowTaskSetSpec{
+				Tasks: []wfv1.Task{
+					{
+						NodeID:   nodeId,
+						Template: tmpl,
+					},
+				},
+			},
+		}
+		err = argowait.Backoff(retry.DefaultBackoff, func() (bool, error) {
+			var err error
+			_, err = wfts.wfTaskSetClient.WorkflowTaskSets(wf.Namespace).Create(ctx, &taskSet, metav1.CreateOptions{})
+			return err == nil, err
+		})
+		if err != nil {
+			log.Errorf(err.Error())
+			return err
+		}
+		return nil
+	} else {
+		taskSet, err := util.UnstructuredToTaskSet(obj)
+		if err != nil {
+			log.WithError(err).Error(fmt.Sprintf("Failed to get TaskSet '%s' from informer index", key))
+			return err
+		}
+		task := wfv1.Task{
+			NodeID:   nodeId,
+			Template: tmpl,
+		}
+		if existing := taskSet.Spec.Tasks.GetTask(nodeId); existing == nil {
+			taskSet.Spec.Tasks = append(taskSet.Spec.Tasks, task)
+			err = argowait.Backoff(retry.DefaultBackoff, func() (bool, error) {
+				var err error
+				_, err = wfts.wfTaskSetClient.WorkflowTaskSets(wf.Namespace).Update(ctx, taskSet, metav1.UpdateOptions{})
+				return err == nil, err
+			})
+
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+func (wfts WorkflowTaskSetManager) Run(ctx context.Context) {
+	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
+	defer wfts.wfTaskSetQueue.ShutDown()
+	log.Infof("Starting TaskSet manager")
+	wfts.wfTaskSetInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(old, new interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(new)
+				if err == nil {
+					wfts.wfTaskSetQueue.Add(key)
```

Comment:
I think it was a mistake to have a pod queue in the controller, it does not add value - instead just add directly to the workflow queue